docs/decisions/eventing-and-messaging.md

```diff
@@ -0,0 +1,67 @@
+# Separating Eventing and Messaging
+
+Decision date: 6 March 2019
+
+## What
+
+Move Channel and Subscription to the apigroup `messaging.knative.dev` from
+`eventing.knative.dev`, to indicate that Channel and Subscription concepts are
+more aligned with messaging (i.e. message routing) than eventing (i.e.
+content-aware dispatch).
+
+## New API Group charters
+
+### `messaging.knative.dev`
+
+Messaging is explicitly concerned with modelling the event transmission
+topology; the route of the message through the topology is specified by the
+developer, and the topology determines which functions receieve which events.
+
+![Channel and Subscription model](channel-subscription-model.svg)
+
+Message Channels can be used for Pub/Sub messaging where producers and consumers
+are decoupled, even temporally: either side can come and go, consumers read at
+their own rate, potentially including support for replay, etc. Channels also
+provide the building blocks for Enterprise Integration Patterns, such as Routers
+and higher level patterns such as Scatter/Gather and may support partitioning,
+and stream processing over windows.
+
+### `eventing.knative.dev`
+
+Eventing is a higher-level abstraction which removes explicit management of the
+event transmission topology and replaces it with automatically-optimized routing
+based on the contents of the event. In this way, it is similar to local
+event-driven programming with the Observer pattern.
+
+![Broker and Trigger model](broker-trigger-model.svg)
+
+An Event Broker is a target for emitting events, and Triggers connect event
+listeners based on selection criteria. This provides the equivalent of
+`emit(foo)` and `on('foo', function(e))` for a distributed system. The
+implementation details are hidden.
+
```

Comment:
Do we want to say anything about the fact that it's possible (and desireable?) to implement eventing on messaging. The original doc had a section on eventing/messaging being complimentary.

Reply:
I don't think it's desirable to implement eventing on Channel and Subscription. It would be better for the implementation details of Broker/Trigger to be completely hidden, whereas the purpose of Channel is to provide a common abstraction across different messaging systems that a user might choose intentionally for certain properties (e.g. support for replalying messages from a given offset). That type of variability of underlying implementation would not be desirable within the Broker/Trigger internals since that could create inconsistency in the behavior.

Reply:
> the purpose of Channel is to provide a common abstraction across different messaging systems 

> That type of variability of underlying implementation would not be desirable within the Broker/Trigger internals since that could create inconsistency in the behavior.

I'm confused by this. Channel is both a common abstraction AND too variable to be relied upon? I thought we had a set of behaviors that all Channels must implement, and that this was the value of the Channel abstraction.

Broker and Trigger may be implemented using pub/sub primitives. In that case, it seems advantageous to share as much code as possible with our existing pub/sub primitives.

Reply:
An abstraction need not be a lowest common denominator. Users will pick Channels based on specific characteristics, eg. QoS. All Channels must provide PubSub semantics for a "consumer group" (per-subscriber) model, but beyond that there is indeed variability based on the underlying middleware.

Broker/Trigger are being described as a black box, and we've heard on WG calls that Google's own product will *not* rely on Channels. It doesn't seem idiomatic to k8s to have some implementations build on lower level CRDs while other implementations don't. For that reason, it would be better to treat Broker/Trigger as a black box without any lower level CRDs being involved.